<a href="https://colab.research.google.com/github/unknown-jun/Basic_of_Python/blob/master/review.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import files
emp = files.upload()

Saving emp3.csv to emp3.csv


In [ ]:
import pandas as pd
emp = pd.read_csv("emp3.csv")

문제 86. 이름이 SCOTT인 사원의 이름과 월급을 출력하시오

In [ ]:
emp[['ename','sal']] [emp['ename']=='SCOTT']

ename   sal
11  SCOTT  3000

문제 89. 판다스를 이용해서 emp3.csv에서 이름을 출력하는데 이름의 첫번째 철자만 출력하시요

In [ ]:
for i in emp['ename']:
    print( i[0] )

K
B
C
J
M
A
T
J
W
F
S
S
A
M


문제 90. 판다스를 이용해서 emp3.csv를 가져와서 이름의 끝글자를 출력하시오

In [ ]:
for i in emp['ename']:
    print( i[-1] )

G
E
K
S
N
N
R
S
D
D
H
T
S
R


문제 91. 위의 결과에서 앞에 이름도 같이 출력하시오

In [ ]:
for i in emp[ 'ename' ]:
    print( i, i[-1] )

KING G
BLAKE E
CLARK K
JONES S
MARTIN N
ALLEN N
TURNER R
JAMES S
WARD D
FORD D
SMITH H
SCOTT T
ADAMS S
MILLER R


문제 92. 이름의 첫번째 철자가 s로 시작하는 사원들의 이름을 출력하시오

In [ ]:
for i in emp['ename']:
    if i[0] == 'S':
        print(i)

SMITH
SCOTT


문제 93. 이름의 끝글자가 T로 끝나는 사원들의 이름을 출력하시오


In [ ]:
for i in emp['ename']:
    if i[-1] == 'T':
        print( i )

SCOTT


문제 94. 이름의 두 번째 철자가 M인 사원의 이름을 출력하시오

In [ ]:
for i in emp[ 'ename' ]:
    if i[ 1 ] == 'M':
        print( i )

SMITH


문제 95. 아래의 SQL을 판다스로 구현하시오
```
> SQL
select substr(ename, 1, 3)
    from emp;
```

In [ ]:
for i in emp[ 'ename' ]:
    print( i[0:3] )

KIN
BLA
CLA
JON
MAR
ALL
TUR
JAM
WAR
FOR
SMI
SCO
ADA
MIL


문제 96. 아래의 SQL을 판다스로 구현하시오
```
> SQL
select substr(ename, -2, 2)
    from emp;
```

In [ ]:
for i in emp[ 'ename' ]:
    print(i[-2:])

NG
KE
RK
ES
IN
EN
ER
ES
RD
RD
TH
TT
MS
ER


문제97. 아래의 SQL을 판다스로 구현하시오
```
> SQL
select substr( ename, 2, 3)
    from emp;
```

In [ ]:
for i in emp['ename']:
    print( i[1:4] )

ING
LAK
LAR
ONE
ART
LLE
URN
AME
ARD
ORD
MIT
COT
DAM
ILL



문제 99. 아래의 SQL을 판다스로 구현하시오
```
> SQL
select ename||sal
    from emp;
```

In [ ]:
for i,k in zip( emp['ename'], emp['sal'] ): # 두개의 변수를 한번에 for loop를 실행할 때 zip을 사용한다.
    print( i + str(k) )
print('\n')
for i,k in zip( emp['ename'], emp['sal'] ):
    print( i, k)

KING5000
BLAKE2850
CLARK2450
JONES2975
MARTIN1250
ALLEN1600
TURNER1500
JAMES950
WARD1250
FORD3000
SMITH800
SCOTT3000
ADAMS1100
MILLER1300


KING 5000
BLAKE 2850
CLARK 2450
JONES 2975
MARTIN 1250
ALLEN 1600
TURNER 1500
JAMES 950
WARD 1250
FORD 3000
SMITH 800
SCOTT 3000
ADAMS 1100
MILLER 1300


문제 100. 주사위의 눈 6개를 100개 담은 리스트 dice100을 만드시오

In [ ]:
dice = [1,2,3,4,5,6]
dice100 = dice *100
print(dice100)

문제 101. 초등학생 키가 10개가 들어있는 아래의 tall 리스트의 요소 10개를 10000개로 증가시켜서 tall10000 리스트에 담고 출력하시오
```
tall = [129.3, 130.2, 132.5, 134.7, 136.3, 137.8, 138.1, 140.2, 142.3, 145.2 ]
```

In [ ]:
tall = [129.3, 130.2, 132.5, 134.7, 136.3, 137.8, 138.1, 140.2, 142.3, 145.2 ]
tall_10000 = tall * 10000

문제 102. 위의 모집단의 평균, 분산, 표준편차 값을 출력하시오

In [ ]:
import numpy as np

print( np.mean( tall_10000 ) )
print( np.var( tall_10000 ))
print( np.std( tall_10000 ))

136.66
23.782399999999985
4.876720209321013



문제 103. 위의 모집단 tall10000에서 표본을 20개를 랜덤으로 추출하시오

In [ ]:
import random as rd

sample = []

for i in range(1, 21):
    random_tall = rd.choice( tall_10000 )
    sample.append(random_tall)
    
print(sample)

[129.3, 134.7, 129.3, 137.8, 138.1, 136.3, 137.8, 136.3, 129.3, 145.2, 137.8, 134.7, 137.8, 136.3, 130.2, 142.3, 136.3, 140.2, 132.5, 137.8]



문제 105. 위의 표본이 평균값을 출력하시오

In [ ]:
np.mean( sample )

136.0

문제 106. 아래의 SQL을 판다스로 구현하시오
```
> SQL
select ename, length(ename)
    from emp;
```

In [ ]:
for i in emp['ename']:
    print( i, len(i) )

KING 4
BLAKE 5
CLARK 5
JONES 5
MARTIN 6
ALLEN 5
TURNER 6
JAMES 5
WARD 4
FORD 4
SMITH 5
SCOTT 5
ADAMS 5
MILLER 6


문제 107. 모평균이 30이고 모표준편차가 7인 모집단을 구성하시오

In [ ]:
avg = 30
std = 7
N = 10000

population = np.random.randn(N) * std + avg
print( len( population ) )

10000



문제 108. 위의 모집단의 모평균을 출력하시오!

In [ ]:
avg = 30
std = 7
N = 10000

population = np.random.randn(N) * std + avg
np.mean( population )

29.99444875872508

문제 109. 아래의 SQL을 판다스로 구현하시오
```
> SQL
select ename, job
    from emp
    where job in ('SALESMAN', 'ANALYST' );
```

In [ ]:
for i,e in zip( emp['ename'], emp['job'] ):
    if e in ( 'SALESMAN', 'ANALYST' ):
        print(i, e)

emp[['ename', 'job']][ emp['job'].isin( ['SALESMAN', 'ANALYST'] ) ]

MARTIN SALESMAN
ALLEN SALESMAN
TURNER SALESMAN
WARD SALESMAN
FORD ANALYST
SCOTT ANALYST


ename       job
4   MARTIN  SALESMAN
5    ALLEN  SALESMAN
6   TURNER  SALESMAN
8     WARD  SALESMAN
9     FORD   ANALYST
11   SCOTT   ANALYST

문제 110. 아래의 SQL을 판다스로 구현하시오
```
> SQL
select ename, job
    from emp
    where job not in ('SALESMAN', 'ANALYST');
```

In [ ]:
emp[['ename', 'job']][~ emp['job'].isin(['SALESMAN', 'ANALYST'])]

ename        job
0     KING  PRESIDENT
1    BLAKE    MANAGER
2    CLARK    MANAGER
3    JONES    MANAGER
7    JAMES      CLERK
10   SMITH      CLERK
12   ADAMS      CLERK
13  MILLER      CLERK


문제 112. 모평균이 30이고 모표준편차가 7인 모집단 1000000개의 모평균과 모분산과 모표준편차를 아래와 같이 출력하시오
```
"모집단"의 모평균은 136.66
"모집단"의 모분산은 23.78239999999999
"모집단"의 모표준편차는 4.876720209321014


In [ ]:
N = 10000
std = 7
avg = 30

population = np.random.randn(N) * std + avg
print(f'"모집단"의 모평균은 {np.mean(population)}' )
print(f'"모집단"의 모분산은 {np.var(population)}' )
print(f'"모집단"의 모표준편차은 {np.std(population)}' )

"모집단"의 모평균은 29.932710987435904
"모집단"의 모분산은 48.44196201886415
"모집단"의 모표준편차은 6.96002600705372



문제 113. 모평균이 30이고 모표준편차가 7인 모집단(1000000개)에서 표본을 49개를 뽑으시오

In [ ]:
sample_list = []

for i in range(49):
    sample = rd.choice(population)
    sample_list.append( sample )

print( sample_list )
np.random.choice( population , 49)

문제 114. 위에서 뽑은 49개의 평균값을 출력하시오

In [ ]:
print( np.mean(sample_list) )
np.mean( np.random.choice( population, 49 ) )

29.275787669775873


29.88238062385161


문제 115. 위에서 출력한 표본이 평균값을 하나가 아니라 100개가 출력되게 하시오

In [ ]:
avg = 30
std = 7
N = 1000000

population = np.random.randn(N) * std + avg
for i in range(1, 101):
    sample = np.random.choice(population, 49)
    print(sample.mean())


문제 116. 이번에는 100개를 출력하지 말고 비어있는 리스트a를 선언하고 a에 100개를 담으시오

In [ ]:
avg = 30
std = 7
N = 10000

population = np.random.randn(N) * std + avg
sample_list = []

for i in range(1, 101):
    sample = np.random.choice(population, 49 )
    sample_list.append( sample )

sample_list

문제 117. 위에서 구한 표본평균의 평균값과 표준편차를 아래와 같이 출력하시오
```
"표본평균" 의 평균값은 ***
"표본평균" 의 표준편차는 ***


In [ ]:
N=10000
std = 7
avg = 30

population = np.random.randn(N) * std + avg
sample_list = []

for i in range(100):
    sample_list.append( np.random.choice(population, 30).mean() )

print(f'"표본평균"의 평균값은 {np.mean(sample_list)}')
print(f'"표본평균"의 표준편차는 {np.std(sample_list)}')

"표본평균"의 평균값은 30.05026870727082
"표본평균"의 표준편차는 1.287740200558459



문제 121. (통계를 코드로 구현 4번) 어느 비행기 탑승객의 짐의 무게는 평균이 18kg이고 표준편차가 3kg인 정규분포를 따른다고 한다. 이 비행기 탑승객 중에서 36명을 임의추출할 때, 짐의 평균 무게가 17kg 이상일 확률을 구하여라

In [ ]:
n = 36
avg = 18
std = 3

population = np.random.randn(N) * std + avg
cnt = 0

for i in range(1, 37):
    sample = np.random.choice( population )
    if sample >= 17:
        cnt += 1

cnt/36

0.6388888888888888


문제 122. 동전의 앞면과 뒷면을 포함하는 리스트 coin을 만드시오

In [ ]:
coin = ['front', 'back']

문제 123. 위에서 만든 coin의 각각의 요소를 100개로 늘려서 coin_100 변수에 담으시오

In [ ]:
coin_100 = coin * 100
len( coin_100 )

200

문제 124. 위에서 만든 coin_100리스트에서 표본을 10개를 추출하시오

In [ ]:
import numpy as np

np.random.choice( coin_100, 10 )

array(['front', 'back', 'front', 'back', 'front', 'back', 'back', 'back',
       'back', 'back'], dtype='<U5')


문제 125. 위에서 추출한 샘플 10개에서 앞면이 몇번 나오는지 출력하시오

In [ ]:
cnt = 0
for i in range(10):
    if np.random.choice( coin_100 ) == 'front':
        cnt += 1

print( cnt, '\n' )

# or

cnt_1 = 0
random_stuff = np.random.choice( coin_100, 10 )

for i in random_stuff:
    if i == 'front':
        cnt_1 += 1

cnt_1

4 



2

문제 129. 위의 튜플의 모든 요소들을 다 출력하시오!
```
결과:
0.01
0.02
0.03
0.04
0.04


In [ ]:
a = (0.01, 0.02, 0.03, 0.04, 0.04)

for i in a:
    print( i )

0.01
0.02
0.03
0.04
0.04


문제 130. 아래의 두개의 리스트를 각각 만들고 아래와 같이 fruit라는 딕셔너리를 생성하시오
```
a=['사과','배', '포도', '복숭아', '바나나']
b=['apple', 'pear', 'grape', 'peach', 'banana'

결과 
print( fruit )
{'사과':'apple', '바나나':'banana', '복숭아':'peach', '배':'pear','포도': 'grape'}
```


In [ ]:
fruit = {}
a=['사과','배', '포도', '복숭아', '바나나']
b=['apple', 'pear', 'grape', 'peach', 'banana']

for i,k in zip(a, b):
    fruit[i] = k

fruit

{'바나나': 'banana', '배': 'pear', '복숭아': 'peach', '사과': 'apple', '포도': 'grape'}

예. 아래의 SQL을 파이썬으로 구현하시오
```
SQL> 
select lower(ename)
    from emp;
    

In [ ]:
for i in emp['ename']:
    print( i.lower() )

king
blake
clark
jones
martin
allen
turner
james
ward
ford
smith
scott
adams
miller



문제 131. 아래의 SQL을 파이썬으로 구현하시오
```
SQL>
select lower(ename), lower(job)
    from emp;
```

In [ ]:
for i, j in zip(emp['ename'], emp['job']):
    print( i.lower(), j.lower() )

king president
blake manager
clark manager
jones manager
martin salesman
allen salesman
turner salesman
james clerk
ward salesman
ford analyst
smith clerk
scott analyst
adams clerk
miller clerk


In [ ]:
# 두개의 입력값을 받아서 더하는 함수를 생성
def sum_num(a,b):
    return (a + b)

sum_num(2, 3)

5

문제 132. for loop문을 사용하여 1부터 10까지 출력하시오

In [ ]:
for i in range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


문제133. 1부터 10까지 다 더한 값을 출력하시오



In [ ]:
cnt = 0

for i in range(1, 11):
    cnt += i

cnt

55

문제 134. 위의 코드를 이용해서 함수를 생성하는데 아래와 같이 숫자를 입력하고 함수를 실행하면 해당 숫자까지 1부터 다 더한 값이 출력되게 하시오!



In [ ]:
def sum_(a):
    cnt = 0
    for i in range(1, a+1):
        cnt += i
    return cnt

sum_(10)

55


문제 135. 다음의 문자열 변수를 생성하고 문자열 변수의 문자를 소문자로 출력하시오
```
a = 'SCOTT'
```

In [ ]:
a = 'SCOTT'
print( 'SCOTT'.lower() )
a.lower()

scott


'scott'


문제 140 아래의 함수를 생성하시오!(첫번째 철자 대문자 나머지 소문자로 출력되게 하는 함수)
```
print( initcap('scott') )
결과: Scott
```

In [ ]:
def capital(a):
    return a[0].upper() + a[1:].lower()

capital('scott')

'Scott'

In [61]:
def add_text(t1, t2):
    return t1 + '  ' + t2

add_text('Python', 'Java')

'Python  Java'

In [65]:
def add_text2(t3, t4='is the best'):
    return t3 + '    ' + t4

print(add_text2('Python'))
add_text2('Python', 'Java')

Python    is the best


'Python    Java'